src/ServiceControl.Monitoring/Timings/TimingsStore.cs

```diff
@@ -3,122 +3,162 @@
+                }
+                finally
+                {
+                    rwl.ExitReadLock();
+                }
+            }
 
-                    result.Add(new EndpointInstanceTimings
+            public void Report(LongValueOccurrences message)
+            {
+                rwl.EnterWriteLock();
+                try
+                {
+                    for (var i = 0; i < message.Length; i++)
                     {
-                        Id = instanceId,
-                        TotalMeasurements = totalMeasurements,
-                        TotalTime = totalDuration,
-                        Intervals = timingIntervals
-                    });
+                        Report(ref message.entries[i]);
+                    }
+                }
+                finally
+                {
+                    rwl.ExitWriteLock();
                 }
             }
 
-            return result.ToArray();
-        }
-
-        static DateTime[] GenerateIntervalIds(DateTime start, int numberOfPastIntervals)
-        {
-            var intervals = new DateTime[numberOfPastIntervals];
+            void Report(ref LongValueOccurrences.Entry entry)
+            {
+                var epoch = GetEpoch(ref entry);
+                var epochIndex = epoch % Size;
 
-            intervals[0] = start.RoundDownToNearest(TimeSpan.FromSeconds(15));
+                if (intervals[epochIndex].Epoch == epoch)
+                {
+                    intervals[epochIndex].TotalTime += entry.Value;
+                    intervals[epochIndex].TotalMeasurements += 1;
+                }
+                else
+                {
+                    // only if epoch is newer than the one written before, overwrite
+                    // this ensures that old, out-of-order messages do not flush the existing data
+                    if (epoch > intervals[epochIndex].Epoch)
+                    {
+                        intervals[epochIndex].Epoch = epoch;
+                        intervals[epochIndex].TotalTime = entry.Value;
+                        intervals[epochIndex].TotalMeasurements = 1;
+                    }
+                }
+            }
 
-            for (var i = 1; i < numberOfPastIntervals; i++)
+            struct MeasurementInterval
             {
-                intervals[i] = intervals[i - 1].Subtract(IntervalSize);
-            }
+                public int TotalMeasurements;
+                public long Epoch;
+                public long TotalTime;
 
-            return intervals;
-        }
+                public override string ToString()
+                {
+                    return $"{nameof(TotalMeasurements)}: {TotalMeasurements}, {nameof(Epoch)}: {Epoch}, {nameof(TotalTime)}: {TotalTime}";
+                }
+            }
 
-        struct MeasurementInterval
-        {
-            public int TotalMeasurements;
-            public long TotalTime;
+            static long GetEpoch(ref LongValueOccurrences.Entry entry)
```

Comment:
The `epochs` are needed only for the `TimingStore`. I see no value in moving this to the `Entry` as `Entry` has no notion of `epoch`.